rpcs3/Emu/RSX/Common/ProgramStateCache.h

```diff
@@ -348,31 +366,40 @@ class program_state_cache
 		// NOTE: Linking is much slower than decompilation step, so always decompile at least 1 unit
 		// TODO: Use try_lock instead
 		bool busy = false;
-		{
-			u32 count = 0;
-			std::lock_guard lock(m_decompiler_mutex);
+		u32 count = 0;
+		std::unique_ptr<async_decompile_task_entry> decompile_task;
 
-			while (!m_decompile_queue.empty())
+		while (true)
+		{
 			{
-				const auto& decompile_task = m_decompile_queue.front();
-				if (decompile_task.is_fp)
+				std::lock_guard lock(m_decompiler_mutex);
+				if (m_decompile_queue.empty())
 				{
-					search_fragment_program(decompile_task.fp);
+					break;
 				}
 				else
 				{
-					search_vertex_program(decompile_task.vp);
+					decompile_task = std::make_unique<async_decompile_task_entry>(std::move(m_decompile_queue.front()));
 				}
+			}
+
+			if (decompile_task->is_fp)
+			{
+				search_fragment_program(decompile_task->fp);
+			}
+			else
+			{
+				search_vertex_program(decompile_task->vp);
+			}
 
-				m_decompile_queue.pop_front();
+			m_decompile_queue.pop_front();
```

Comment:
All container access has to be guarded for implementation reasons; move this after https://github.com/RPCS3/rpcs3/pull/7190/files#diff-100f7e09dac2a73d3b7e90d5497d4a28R382 so that it happens under the lock.